Plugin.cs

```diff
@@ -255,5 +255,75 @@ public void ReplacePlayer(Player playerToReplace)
             });
 
         }
+        
+        private void OnLeft(LeftEventArgs ev)
+        {
+            if(ev.Player.IsDead)
+                return;
+
+            List<Player> spectators = ListPool<Player>.Pool.Get(Player.Get(RoleTypeId.Spectator));
+
+            if (spectators.IsEmpty())
+            {
+                ListPool<Player>.Pool.Return(spectators);
+                return;
+            }
+        
+            Player chosenSpectator = spectators.RandomItem();
+            
+            foreach (CustomRole customRole in ev.Player.GetCustomRoles())
+            {
+                customRole.AddRole(chosenSpectator);
+            }
+            
+            if(chosenSpectator.IsDead)
+                chosenSpectator.Role.Set(ev.Player.Role.Type);
+            
+            chosenSpectator.Teleport(ev.Player.Position);
+        
+            chosenSpectator.MaxHealth = ev.Player.MaxHealth;
+            chosenSpectator.Health = ev.Player.Health;
+            chosenSpectator.MaxArtificialHealth = ev.Player.MaxArtificialHealth;
+            chosenSpectator.ArtificialHealth = ev.Player.ArtificialHealth;
+            chosenSpectator.HumeShield = ev.Player.HumeShield;
+        
+            if (ev.Player.Role.Is(out Scp079Role scp))
+            {
+                float playerEnergy = scp.Energy;
+                int playerXp = scp.Experience;
+                int playerLevel = scp.Level;
+                Camera playerCamera = scp.Camera;
+                float playerBlackoutCoolDown = scp.BlackoutZoneCooldown;
+            
+                if (chosenSpectator.Role.Is(out Scp079Role newScp))
+                {
+                    newScp.Energy = playerEnergy;
+                    newScp.Experience = playerXp;
+                    newScp.Level = playerLevel;
+                    newScp.Camera = playerCamera;
+                    newScp.BlackoutZoneCooldown = playerBlackoutCoolDown;
+                }
+            }
+
+            Timing.CallDelayed(0.3f, () => chosenSpectator.ClearInventory());
+        
+            foreach (Item item in ev.Player.Items.ToArray())
+            {
+                CustomItem.TryGet(item, out CustomItem? customItem);
+                chosenSpectator.GiveItemDelayed(item.Clone(), customItem);
+                item.Destroy();
+            }
+        
+            foreach (var itemType in ev.Player.Ammo)
+            {
+                Timing.CallDelayed(0.4f, () => chosenSpectator.SetAmmo(itemType.Key.GetAmmoType(), itemType.Value));
+            }
+        
```

Comment:
Most of this code is the same as the other replace function; it would be best to pull out common functionality to a shared function rather than copy-paste so that it's easier to maintain

Reply:
broke down repeated code into a seperate shared function